attribution/api_attribution.py

```diff
@@ -46,11 +52,244 @@ def __init__(
+                original_input,
+                original_output.message.content,
+                perturbation_strategy,
+                perturb_word_wise,
+            )
+
+        units, tokens_per_unit, ids_per_unit = get_units_from_prompt(original_input, self.tokenizer, perturb_word_wise=perturb_word_wise)
+
+        perturbations: list[PerturbedLLMInput] = []
+        for i_unit, unit in enumerate(units):
+            perturbed_unit = [
+                self.tokenizer.decode(perturbation_strategy.get_replacement_token(token_id)).strip()
+                for token_id in ids_per_unit[i_unit]
+            ]
+
+            perturbed_input = deepcopy(tokens_per_unit)
+            perturbed_input[i_unit] = perturbed_unit
+
+            perturbation = PerturbedLLMInput(
+                input_units=perturbed_input, 
+                masked_units=[unit], 
+                unit_idx=[i_unit], 
+                tokenizer=self.tokenizer,
+            )
+            perturbations.append(perturbation)
+
+        outputs = await self.get_multiple_completions([perturbation.input_string for perturbation in perturbations])
+        
+        for perturbation, output in zip(perturbations, outputs):
+            for strategy in attribution_strategies:    
+                attribution_scores, _ = self._get_scores(
+                    perturbed_output=output, 
+                    original_output=original_output, 
+                    attribution_strategy=strategy,
+                    ignore_output_token_location=ignore_output_token_location
+                )
+
+                if logger:
+                    logger.log_attributions(
+                        perturbation, 
+                        attribution_scores,
+                        strategy,
+                        output.message.content
+                    )
+
+                    logger.log_perturbation(
+                        len(perturbation.masked_units) - 1,
+                        perturbation.masked_string,
+                        str(perturbation_strategy),
+                        original_input,
+                        original_output.message.content,
+                        perturbation.input_string,
+                        output.message.content,
+                    )
+            
+        if logger:
+            logger.stop_experiment(num_llm_calls=len(outputs) + 1)
+
+    async def hierarchical_perturbation(
+            self, 
+            original_input: str, 
+            init_chunk_size: int,
+            threshold: float = 0.5, 
+            stride: Optional[int] = None, 
+            logger: Optional[ExperimentLogger] = None, 
+            perturbation_strategy: PerturbationStrategy = FixedPerturbationStrategy(),
+            attribution_strategies: list[str] = ["cosine", "prob_diff"],
+            perturb_word_wise: bool = False, 
+            ignore_output_token_location: bool = True,
+            verbose: bool = False,
+        ) -> pd.Series:
```

Comment:
Do you think it makes sense to move sections of this algorithm to smaller functions? It may help for readibility. I'll highlight some sections on comments below for potential candidates.

Reply:
I've split up the function now, pretty much in the same way you suggested!